File: mount_scripts/explore_phages.py
"""
Python Script: PBI-Scope Database Exploration

This script connects to the PBI-Scope database using the pbi package
and generates exploration plots. It demonstrates how to use Python
with the PBI-Scope data.

Usage:
    python explore_phages.py

Prerequisites:
    - Python packages: pbi, duckdb, pandas, matplotlib, seaborn
    - The pbi-data Docker volume must be mounted at /data

Output:
    Plots are saved to ./output/ directory

This script is designed to run inside the custom R+Python container.
See Dockerfile and docker-compose.custom.yml for setup instructions.
"""

import os
from pathlib import Path

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# The pbi package provides quick_connect() for easy database access.
# It auto-detects the DATA_PATH environment variable and connects
# to the database and FASTA files automatically.
from pbi import quick_connect


def main():
    """Run the exploration analysis."""
    print("=== PBI-Scope Database Exploration (Python) ===\n")

    # Create output directory
    output_dir = Path("output")
    output_dir.mkdir(exist_ok=True)

    # Connect to database
    # quick_connect() reads DATA_PATH env var and resolves all file paths.
    # In Docker: DATA_PATH=/data/processed
    print("Connecting to database...")
    retriever = quick_connect()

    # --- Plot 1: Source Database Distribution ---------------------------------
    # Which public databases contribute the most phages?
    print("[1/6] Querying source database distribution...")
    source_dist = retriever.conn.execute("""
        SELECT Source_DB, COUNT(*) AS phage_count
        FROM fact_phages
        GROUP BY Source_DB
        ORDER BY phage_count DESC
    """).fetchdf()

    print("  Top sources:")
    print(source_dist.head().to_string(index=False))

    fig, ax = plt.subplots(figsize=(10, 6))
    source_dist.plot(
        kind="barh", x="Source_DB", y="phage_count", 
        ax=ax, legend=False, color="steelblue"
    )
    ax.set_title(f"Phages by Source Database\nTotal: {source_dist['phage_count'].sum():,} phages")
    ax.set_xlabel("Number of Phages")
    ax.set_ylabel(None)
    ax.xaxis.set_major_formatter(plt.FuncFormatter(lambda x, p: f"{x:,.0f}"))
    plt.tight_layout()
    fig.savefig(output_dir / "01_source_distribution.png", dpi=150)
    plt.close(fig)
    print("  Saved: output/01_source_distribution.png\n")

    # --- Plot 2: Genome Length Distribution -----------------------------------
    # Phage genome lengths vary widely (from ~2 kb to ~500 kb).
    print("[2/6] Querying genome length distribution...")
    lengths = retriever.conn.execute("""
        SELECT Length
        FROM fact_phages
        WHERE Length > 0 AND Length < 200000
    """).fetchdf()

    median_len = lengths["Length"].median()
    print(f"  Median: {median_len:,.0f} bp")
    print(f"  Mean: {lengths['Length'].mean():,.0f} bp")
    print(f"  Min: {lengths['Length'].min():,.0f} bp")
    print(f"  Max: {lengths['Length'].max():,.0f} bp\n")

    fig, ax = plt.subplots(figsize=(10, 6))
    lengths["Length"].div(1000).hist(
        bins=50, ax=ax, color="steelblue", edgecolor="white"
    )
    ax.axvline(median_len / 1000, color="red", linestyle="--", 
               linewidth=1.5, label=f"Median: {median_len/1000:,.0f} kb")
    ax.set_title("Phage Genome Length Distribution")
    ax.set_xlabel("Genome Length (kb)")
    ax.set_ylabel("Count")
    ax.legend()
    plt.tight_layout()
    fig.savefig(output_dir / "02_length_distribution.png", dpi=150)
    plt.close(fig)
    print("  Saved: output/02_length_distribution.png\n")

    # --- Plot 3: Lifestyle Distribution ---------------------------------------
    # Phages are classified as virulent (lytic) or temperate (lysogenic).
    print("[3/6] Querying lifestyle distribution...")
    lifestyle = retriever.conn.execute("""
        SELECT Lifestyle, COUNT(*) AS count
        FROM fact_phages
        WHERE Lifestyle IN ('virulent', 'temperate')
        GROUP BY Lifestyle
    """).fetchdf()

    total = lifestyle["count"].sum()
    for _, row in lifestyle.iterrows():
        pct = row["count"] / total * 100
        print(f"  {row['Lifestyle']}: {row['count']:,} ({pct:.1f}%)")

    fig, ax = plt.subplots(figsize=(8, 6))
    colors = {"virulent": "#E74C3C", "temperate": "#3498DB"}
    lifestyle.plot(
        kind="bar", x="Lifestyle", y="count", ax=ax, legend=False,
        color=[colors.get(x, "gray") for x in lifestyle["Lifestyle"]]
    )
    ax.set_title("Phage Lifestyle Distribution")
    ax.set_xlabel(None)
    ax.set_ylabel("Count")
    ax.yaxis.set_major_formatter(plt.FuncFormatter(lambda x, p: f"{x:,.0f}"))
    plt.xticks(rotation=0)
    plt.tight_layout()
    fig.savefig(output_dir / "03_lifestyle_distribution.png", dpi=150)
    plt.close(fig)
    print("  Saved: output/03_lifestyle_distribution.png\n")

    # --- Plot 4: Top Host Species ---------------------------------------------
    # Which bacterial species have the most known phages?
    print("[4/6] Querying top host species...")
    hosts = retriever.conn.execute("""
        SELECT h.Species_Name, COUNT(DISTINCT p.Phage_ID) AS phage_count
        FROM phage_host_associations pha
        JOIN fact_phages p ON pha.Phage_ID = p.Phage_ID
        JOIN dim_hosts h ON pha.Host_ID = h.Host_ID
        WHERE h.Species_Name IS NOT NULL
        GROUP BY h.Species_Name
        ORDER BY phage_count DESC
        LIMIT 20
    """).fetchdf()

    print("  Top 5 hosts:")
    print(hosts.head().to_string(index=False))

    fig, ax = plt.subplots(figsize=(10, 8))
    hosts.plot(
        kind="barh", x="Species_Name", y="phage_count", 
        ax=ax, legend=False, color="forestgreen"
    )
    ax.set_title("Top 20 Host Species by Number of Phages")
    ax.set_xlabel("Number of Phages")
    ax.set_ylabel(None)
    ax.xaxis.set_major_formatter(plt.FuncFormatter(lambda x, p: f"{x:,.0f}"))
    plt.tight_layout()
    fig.savefig(output_dir / "04_top_hosts.png", dpi=150)
    plt.close(fig)
    print("  Saved: output/04_top_hosts.png\n")

    # --- Plot 5: GC Content Comparison ----------------------------------------
    # Comparing GC content between phages and their hosts.
    print("[5/6] Querying GC content comparison...")
    gc_comparison = retriever.conn.execute("""
        SELECT 
            'Phage' AS type,
            GC_content AS GC
        FROM fact_phages
        WHERE GC_content > 0 AND GC_content < 100
        UNION ALL
        SELECT
            'Host' AS type,
            GC_Content AS GC
        FROM dim_hosts
        WHERE GC_Content > 0 AND GC_Content < 100
    """).fetchdf()

    for t in ["Phage", "Host"]:
        subset = gc_comparison[gc_comparison["type"] == t]
        print(f"  {t} GC: mean={subset['GC'].mean():.1f}%, median={subset['GC'].median():.1f}%")

    fig, ax = plt.subplots(figsize=(10, 6))
    for t, color in [("Phage", "#E74C3C"), ("Host", "#3498DB")]:
        subset = gc_comparison[gc_comparison["type"] == t]
        subset["GC"].plot.kde(ax=ax, color=color, alpha=0.5, label=t, linewidth=2)
    ax.set_title("GC Content Distribution: Phages vs Hosts")
    ax.set_xlabel("GC Content (%)")
    ax.set_ylabel("Density")
    ax.legend()
    plt.tight_layout()
    fig.savefig(output_dir / "05_gc_content_comparison.png", dpi=150)
    plt.close(fig)
    print("  Saved: output/05_gc_content_comparison.png\n")

    # --- Plot 6: Host Assembly Quality ----------------------------------------
    # The quality of host assemblies affects downstream analyses.
    print("[6/6] Querying host assembly quality...")
    assembly = retriever.conn.execute("""
        SELECT Assembly_Level, COUNT(*) AS count
        FROM dim_hosts
        GROUP BY Assembly_Level
        ORDER BY count DESC
    """).fetchdf()

    for _, row in assembly.iterrows():
        pct = row["count"] / assembly["count"].sum() * 100
        print(f"  {row['Assembly_Level']}: {row['count']:,} ({pct:.1f}%)")

    # Order by quality level
    level_order = ["Complete Genome", "Chromosome", "Scaffold", "Contig"]
    assembly["Assembly_Level"] = pd.Categorical(
        assembly["Assembly_Level"], categories=level_order, ordered=True
    )
    assembly = assembly.sort_values("Assembly_Level")

    fig, ax = plt.subplots(figsize=(10, 6))
    assembly.plot(
        kind="bar", x="Assembly_Level", y="count", ax=ax, legend=False,
        color=plt.cm.Set2(range(len(assembly)))
    )
    ax.set_title("Host Assembly Quality Distribution")
    ax.set_xlabel(None)
    ax.set_ylabel("Number of Hosts")
    ax.yaxis.set_major_formatter(plt.FuncFormatter(lambda x, p: f"{x:,.0f}"))
    plt.xticks(rotation=45, ha="right")
    plt.tight_layout()
    fig.savefig(output_dir / "06_assembly_quality.png", dpi=150)
    plt.close(fig)
    print("  Saved: output/06_assembly_quality.png\n")

    # Cleanup
    retriever.close()

    print("=== All plots saved to output/ ===")
    print("Files:")
    for f in sorted(output_dir.glob("*.png")):
        print(f"  - {f.name}")


if __name__ == "__main__":
    main()

File: mount_scripts/explore_phages.R
# =============================================================================
# R Script: PBI-Scope Database Exploration
# =============================================================================
# This script connects to the PBI-Scope DuckDB database and generates
# exploration plots. It demonstrates how to use R with the PBI-Scope data.
#
# Usage:
#   Rscript explore_phages.R
#
# Prerequisites:
#   - R packages: DBI, duckdb, dplyr, ggplot2, scales
#   - The pbi-data Docker volume must be mounted at /data
#
# Output:
#   Plots are saved to ./output/ directory
#
# This script is designed to run inside the custom R+Python container.
# See Dockerfile and docker-compose.custom.yml for setup instructions.
# =============================================================================

# --- Load Libraries -----------------------------------------------------------
# DBI: Database Interface - standard R interface for databases
# duckdb: DuckDB driver for DBI
# dplyr: Data manipulation grammar
# ggplot2: Grammar of graphics for visualization
# scales: Axis formatting utilities (comma labels, percent, etc.)
library(DBI)
library(duckdb)
library(dplyr)
library(ggplot2)
library(scales)

# --- Configuration ------------------------------------------------------------
# The database path is determined by the DATA_PATH environment variable.
# In Docker: DATA_PATH=/data/processed (set in docker-compose.yml)
# The database file is at: /data/processed/databases/phage_database_optimized.duckdb
db_path <- file.path(Sys.getenv("DATA_PATH", "/data/processed"),
                     "databases", "phage_database_optimized.duckdb")

# Output directory for plots
output_dir <- "output"
dir.create(output_dir, showWarnings = FALSE)

cat("=== PBI-Scope Database Exploration (R) ===\n\n")
cat("Database:", db_path, "\n")
cat("Output:", output_dir, "/\n\n")

# --- Connect to Database ------------------------------------------------------
# We open a read-only connection to prevent accidental modifications.
cat("Connecting to database...\n")
con <- dbConnect(duckdb(), dbdir = db_path, read_only = TRUE)

# List available tables
tables <- dbListTables(con)
cat("Available tables:", paste(tables, collapse = ", "), "\n\n")

# --- Query 1: Source Database Distribution ------------------------------------
# Which public databases contribute the most phages?
# This helps understand data composition and potential biases.
cat("[1/6] Querying source database distribution...\n")
source_dist <- dbGetQuery(con, "
    SELECT Source_DB, COUNT(*) AS phage_count
    FROM fact_phages
    GROUP BY Source_DB
    ORDER BY phage_count DESC
")

cat("  Top sources:\n")
head(source_dist, 5) %>% 
    mutate(phage_count = comma(phage_count)) %>%
    print()

p1 <- ggplot(source_dist, aes(x = reorder(Source_DB, phage_count), y = phage_count)) +
    geom_bar(stat = "identity", fill = "steelblue") +
    coord_flip() +
    scale_y_continuous(labels = comma) +
    labs(
        title = "Phages by Source Database",
        subtitle = paste("Total:", comma(sum(source_dist$phage_count)), "phages"),
        x = NULL,
        y = "Number of Phages"
    ) +
    theme_minimal(base_size = 12)

ggsave(file.path(output_dir, "01_source_distribution.png"), p1, 
       width = 10, height = 6, dpi = 150)
cat("  Saved: output/01_source_distribution.png\n\n")

# --- Query 2: Genome Length Distribution -------------------------------------
# Phage genome lengths vary widely (from ~2 kb to ~500 kb).
# This distribution helps identify unusual phages and set filtering thresholds.
cat("[2/6] Querying genome length distribution...\n")
lengths <- dbGetQuery(con, "
    SELECT Length
    FROM fact_phages
    WHERE Length > 0 AND Length < 200000
")

cat("  Summary:\n")
cat("    Median:", comma(round(median(lengths$Length))), "bp\n")
cat("    Mean:", comma(round(mean(lengths$Length))), "bp\n")
cat("    Min:", comma(min(lengths$Length)), "bp\n")
cat("    Max:", comma(max(lengths$Length)), "bp\n\n")

p2 <- ggplot(lengths, aes(x = Length / 1000)) +
    geom_histogram(bins = 50, fill = "steelblue", color = "white") +
    geom_vline(xintercept = median(lengths$Length) / 1000, 
               linetype = "dashed", color = "red", linewidth = 1) +
    scale_x_continuous(labels = comma) +
    labs(
        title = "Phage Genome Length Distribution",
        subtitle = paste("Median:", comma(round(median(lengths$Length) / 1000)), "kb"),
        x = "Genome Length (kb)",
        y = "Count"
    ) +
    theme_minimal(base_size = 12)

ggsave(file.path(output_dir, "02_length_distribution.png"), p2, 
       width = 10, height = 6, dpi = 150)
cat("  Saved: output/02_length_distribution.png\n\n")

# --- Query 3: Lifestyle Distribution -----------------------------------------
# Phages are classified as virulent (lytic) or temperate (lysogenic).
# This is a key biological variable for many analyses.
cat("[3/6] Querying lifestyle distribution...\n")
lifestyle <- dbGetQuery(con, "
    SELECT Lifestyle, COUNT(*) AS count
    FROM fact_phages
    WHERE Lifestyle IN ('virulent', 'temperate')
    GROUP BY Lifestyle
")

cat("  Distribution:\n")
lifestyle %>%
    mutate(percentage = paste0(round(count / sum(count) * 100, 1), "%")) %>%
    print()

p3 <- ggplot(lifestyle, aes(x = Lifestyle, y = count, fill = Lifestyle)) +
    geom_bar(stat = "identity") +
    scale_y_continuous(labels = comma) +
    scale_fill_manual(values = c("virulent" = "#E74C3C", "temperate" = "#3498DB")) +
    labs(
        title = "Phage Lifestyle Distribution",
        x = NULL,
        y = "Count"
    ) +
    theme_minimal(base_size = 12) +
    theme(legend.position = "none")

ggsave(file.path(output_dir, "03_lifestyle_distribution.png"), p3, 
       width = 8, height = 6, dpi = 150)
cat("  Saved: output/03_lifestyle_distribution.png\n\n")

# --- Query 4: Top Host Species ------------------------------------------------
# Which bacterial species have the most known phages?
# This reveals sampling bias in phage research.
cat("[4/6] Querying top host species...\n")
hosts <- dbGetQuery(con, "
    SELECT h.Species_Name, COUNT(DISTINCT p.Phage_ID) AS phage_count
    FROM phage_host_associations pha
    JOIN fact_phages p ON pha.Phage_ID = p.Phage_ID
    JOIN dim_hosts h ON pha.Host_ID = h.Host_ID
    WHERE h.Species_Name IS NOT NULL
    GROUP BY h.Species_Name
    ORDER BY phage_count DESC
    LIMIT 20
")

cat("  Top 5 hosts:\n")
head(hosts, 5) %>%
    mutate(phage_count = comma(phage_count)) %>%
    print()

p4 <- ggplot(hosts, aes(x = reorder(Species_Name, phage_count), y = phage_count)) +
    geom_bar(stat = "identity", fill = "forestgreen") +
    coord_flip() +
    scale_y_continuous(labels = comma) +
    labs(
        title = "Top 20 Host Species by Number of Phages",
        x = NULL,
        y = "Number of Phages"
    ) +
    theme_minimal(base_size = 12)

ggsave(file.path(output_dir, "04_top_hosts.png"), p4, 
       width = 10, height = 8, dpi = 150)
cat("  Saved: output/04_top_hosts.png\n\n")

# --- Query 5: GC Content Comparison ------------------------------------------
# Comparing GC content between phages and their hosts can reveal
# co-evolution patterns and horizontal gene transfer.
cat("[5/6] Querying GC content comparison...\n")
gc_comparison <- dbGetQuery(con, "
    SELECT 
        'Phage' AS type,
        GC_content AS GC
    FROM fact_phages
    WHERE GC_content > 0 AND GC_content < 100
    UNION ALL
    SELECT
        'Host' AS type,
        GC_Content AS GC
    FROM dim_hosts
    WHERE GC_Content > 0 AND GC_Content < 100
")

cat("  Summary:\n")
gc_comparison %>%
    group_by(type) %>%
    summarise(
        mean_gc = round(mean(GC), 1),
        median_gc = round(median(GC), 1),
        .groups = "drop"
    ) %>%
    print()

p5 <- ggplot(gc_comparison, aes(x = GC, fill = type)) +
    geom_density(alpha = 0.5) +
    scale_fill_manual(values = c("Phage" = "#E74C3C", "Host" = "#3498DB")) +
    labs(
        title = "GC Content Distribution: Phages vs Hosts",
        x = "GC Content (%)",
        y = "Density",
        fill = "Type"
    ) +
    theme_minimal(base_size = 12)

ggsave(file.path(output_dir, "05_gc_content_comparison.png"), p5, 
       width = 10, height = 6, dpi = 150)
cat("  Saved: output/05_gc_content_comparison.png\n\n")

# --- Query 6: Host Assembly Quality -------------------------------------------
# The quality of host assemblies affects downstream analyses.
# This plot shows the distribution of assembly levels.
cat("[6/6] Querying host assembly quality...\n")
assembly <- dbGetQuery(con, "
    SELECT Assembly_Level, COUNT(*) AS count
    FROM dim_hosts
    GROUP BY Assembly_Level
    ORDER BY count DESC
")

cat("  Distribution:\n")
assembly %>%
    mutate(percentage = paste0(round(count / sum(count) * 100, 1), "%")) %>%
    print()

# Order assembly levels by quality
assembly$Assembly_Level <- factor(assembly$Assembly_Level, 
    levels = c("Complete Genome", "Chromosome", "Scaffold", "Contig"))

p6 <- ggplot(assembly, aes(x = Assembly_Level, y = count, fill = Assembly_Level)) +
    geom_bar(stat = "identity") +
    scale_y_continuous(labels = comma) +
    scale_fill_brewer(palette = "Set2") +
    labs(
        title = "Host Assembly Quality Distribution",
        x = NULL,
        y = "Number of Hosts"
    ) +
    theme_minimal(base_size = 12) +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(file.path(output_dir, "06_assembly_quality.png"), p6, 
       width = 10, height = 6, dpi = 150)
cat("  Saved: output/06_assembly_quality.png\n\n")

# --- Disconnect ---------------------------------------------------------------
dbDisconnect(con, shutdown = TRUE)

cat("=== All plots saved to output/ ===\n")
cat("Files:\n")
list.files(output_dir, pattern = "\\.png$") %>%
    paste0("  - ", .) %>%
    cat(sep = "\n")
